### PR TITLE
Support Z Loss in CE

### LINKED ISSUE #197: Z Loss in CE
### 🚀 The feature, motivation and pitch

Often used in pretraining of LMs for stabilization, i.e. the recent [Chameleon](https://arxiv.org/abs/2405.09818) & [PaLM](https://www.jmlr.org/papers/v24/22-1144.html).

### Alternatives

[flash-attn](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/ops/triton/cross_entropy.py) has implementations of abovementioned features, however, does not support fusing with linear head.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Tcc0403**
## Summary
This PR aims to resolve #197

Implemented z loss in LigerCrossEntropy.

note: `lse_square_scale` not exposed at flce yet, because lse calculation will be different if the loss is computed chunk by chunk
### For loss:
```math
\begin{align}
L_{total} &= L_{ce} + z\_loss\
z\_loss &= lse\_square\_scale \cdot lse^2\
lse &= log \sum e^{X_i}
\end{align}
```
We can use  $m = max(X_i)$ and $d = \sum e^{X_i - m}$, obtained from online softmax algorithm, to calculate $lse$ directly.
```math
\begin{align}
lse &= log \sum e^{X_i}\
     &= log \sum e^{X_i - m + m} = log \sum e^{X_i -m} \cdot e^m\
     &= log\ e^m\sum e^{X_i - m} = m + d
\end{align}
```
### For gradients:
First, we calculate the derivative of lse
```math
\begin{align}
\frac{\partial}{\partial x_i}(lse) &= \frac{\partial}{\partial x_i}(log \sum e^{x_i}) \
                                           &= \frac{1}{\sum e^{x_i}} \cdot  \frac{\partial}{\partial x_i} \sum e^{x_i}\
                                           &= \frac{e^{x_i}}{\sum e^{x_i}} = softmax(x_i).
\end{align}
```
Then we can obtain the derivative of z_loss by chain rule.
```math
\frac{\partial z\_loss}{\partial x_i} = \frac{\partial}{\partial x_i}\left( lse\_square\_scale \cdot lse^2\right)  = 2\cdot lse\_square\_scale \cdot lse \cdot  softmax(x_i),
```
and we have the derivative of cross entropy loss with label smoothing
```math
\frac{\partial L_{ce}}{\partial x_i} = softmax(x_i) - (1 - \epsilon)\delta_{k,y} + \frac{\epsilon}{K}= \begin{cases} softmax(x_i) - \frac{\epsilon}{K},                        &  i \neq y \\
                                                   softmax(x_i) - \frac{\epsilon}{K} - (1 - \epsilon) &  i = y \end{cases}
```
where $\epsilon$ is label_smoothing and $K$ is the number of total classes.
Thus, the derivative of total loss is
```math
\begin{align}
\frac{\partial}{\partial x_i}L_{total} &= \frac{\partial}{\partial x_i}L_{ce} + \frac{\partial}{\partial x_i}z\_loss\
                                                     &= softmax(x_i) - \frac{\epsilon}{K} - (1 - \epsilon)\delta_{k,y} +  2\cdot lse\_square\_scale \cdot lse \cdot softmax(x_i)\
                                                     &=\begin{cases} (1 + 2\cdot lse\_square\_scale \cdot lse)\ softmax(x_i) - \frac{\epsilon}{K}, & i \neq y\\
(1 + 2\cdot lse\_square\_scale \cdot lse)\ softmax(x_i) - \frac{\epsilon}{K} -  (1 - \epsilon), & i = y \end{cases} 
\end{align}
```
### Reference
[PaLM: Scaling Language Modeling with Pathways](https://www.jmlr.org/papers/v24/22-1144.html)
[Chameleon: Mixed-Modal Early-Fusion Foundation Models](https://arxiv.org/abs/2405.09818)
## Testing Done
[benchmark gist](https://gist.github.com/Tcc0403/b9120282334196f66b5169d9f52bccaa)
neglectable error in speed benchmark. 

This benchmark was done on my machine, which is probably not accurate.
```
liger ce: 66.123ms
Peak mem:  8.66200832

liger ce with zloss: 65.991ms
Peak mem:  8.66200832

liger ce with zloss with return zloss: 65.951ms
Peak mem:  8.662073856
```

- Hardware Type: <BLANK>
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: src/liger_kernel/ops/cross_entropy.py [hunk 1]
```diff
@@ -10,11 +13,14 @@ def liger_cross_entropy_kernel(
     Y_ptr,
     Y_stride,
     loss_ptr,
+    z_loss_ptr,
     loss_stride,
     n_cols,
     n_non_ignore,
     ignore_index,
     label_smoothing: tl.constexpr,
+    z_loss_scale: tl.constexpr,
```
Comment (Tcc0403) [inline src/liger_kernel/ops/cross_entropy.py:22]: I'm not sure if making label_smoothing and z_loss_scale `tl.constexpr` is a correct move.
Not familiar with model training. Are these two parameters often changed in practice? I'm worried that it might cause the same issue as #146.

[Flash-attention's implementation](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/ops/triton/cross_entropy.py) creates a new constexpr for it in `triton.heuristics` to solve branching issues. 
I wonder what the difference is between 
1) declare`label_smoothing` as a constexpr, and
2) do calculations in `triton.heuristics` then assign a value to the constexpr `HAS_SMOOTHING`

My assumption is that:
in case 1, JIT every time `label_smoothing` changes
in case 2, JIT only when  `HAS_SMOOTHING` changes because of calculations on `label_smoothing`.

If so, I will go with flash-attn's approach.

File: src/liger_kernel/transformers/cross_entropy.py [hunk 2]
```diff
@@ -1,16 +1,34 @@
-from torch.nn import CrossEntropyLoss
+import torch.nn as nn
 
 from liger_kernel.ops.cross_entropy import LigerCrossEntropyFunction
 
 
-class LigerCrossEntropyLoss(CrossEntropyLoss):
-    def __init__(self, *args, **kwargs):
-        super(LigerCrossEntropyLoss, self).__init__(*args, **kwargs)
```
Comment (Tcc0403) [inline src/liger_kernel/transformers/cross_entropy.py:8]: torch's CrossEntropyLoss doesn't have z_loss parameters

File: src/liger_kernel/ops/fused_linear_cross_entropy.py [hunk 3]
```diff
@@ -189,6 +204,7 @@ def forward(
         weight: (V, H) where V is the number of classes
         bias: (V) where V is the number of classes
         ignore_index: the index to ignore in the target
+        label_smoothing (float): The amount of smoothing when computing the loss, where 0.0 means no smoothing.
```
Comment (Tcc0403) [inline src/liger_kernel/ops/fused_linear_cross_entropy.py:207]: forgot to add in label_smoothing's pr, it can be a single issue.

**Comment by Tcc0403** [comment]
Rename z_loss_scale to lse_square_scale because lse_square_scale is more accurate.
